KAFKA-8982: Add retry of fetching metadata to Admin.deleteRecords

tinaselenge opened the pull request:
Use AdminApiDriver class to refresh the metadata and retry the request that failed with retriable errors. 
Modified the unit and integration test cases for LEADER_NOT_AVAILABLE or NOT_LEADER_OR_FOLLOWER as they are now retried using the driver. 



### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tinaselenge committed:
KAFKA-8982: Add retry of fetching metadata to Admin.deleteRecords

divijvaidya commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,170 @@
+        List<TopicPartition> unmapped,
+        Set<TopicPartition> retriable
+    ) {
+        if (error == Errors.NOT_LEADER_OR_FOLLOWER || error == Errors.LEADER_NOT_AVAILABLE) {
```
Please extract out handling of NOT_LEADER_OR_FOLLOWER at a common place for clients: e.g. other usage https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java#L158

divijvaidya commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,170 @@
+        List<TopicPartition> unmapped,
+        Set<TopicPartition> retriable
+    ) {
+        if (error == Errors.NOT_LEADER_OR_FOLLOWER || error == Errors.LEADER_NOT_AVAILABLE) {
```
instead, could we use `error.exception() instanceof InvalidMetadataException`? That covers all cases where metadata might be a problem. e.g. the current code is missing `UNKNOWN_TOPIC_OR_PARTITION`

dimitarndimitrov replied:
@divijvaidya FWIW when the somewhat similar process was being applied to the `ListOffsets` API and I had the exact same thought, @dengziming noted that not all types of `InvalidMetadataException` should result in a retry (e.g. `KafkaStorageException`). See https://github.com/apache/kafka/pull/13432#discussion_r1167756758

divijvaidya commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,170 @@
+            retriable.add(topicPartition);
+        } else {
+            log.error(
+                "DeleteRecords request for topic partition {} failed due to an unexpected error {}",
```
we should probably catch `TOPIC_AUTHORIZATION_FAILED` separately since that is not "unexpected"

divijvaidya reviewed (commented):
Nice change! Added some comments on error handling. Otherwise looks good.

showuon commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -2291,6 +2289,8 @@ public void testDeleteRecords() throws Exception {
         try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(cluster)) {
             env.kafkaClient().setNodeApiVersions(NodeApiVersions.create());
 
+            env.kafkaClient().prepareResponse(prepareMetadataResponse(cluster, Errors.NONE));
```
Before looking into the implementation, I don't think this test is testing what we expected. The original comment is commenting around this lines:
https://github.com/apache/kafka/pull/7296/files#diff-5422d10d9a7f4776c6538ae3aea27f24e94cf4ecf5e752040125aca6edc795d3R3671-R3682

And it said, when metadata response (mr) contains error, we just fail the future without retry. 

And here, the existing metadataResponse also fail without retry: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L2969-L2972

The point is, we want to retry `metadataResponse`, not `deleteRecordResponse`. What I expected is tests like this:

https://github.com/apache/kafka/blob/513e1c641d63c5e15144f9fcdafa1b56c5e5ba09/clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java#L5431C11-L5433

Please update the test and make sure it passed. 
Thanks.

showuon reviewed (commented):
@tinaselenge , left a meta-comment. Thanks.

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,170 @@
+                    .setPartitionIndex(topicPartition.partition())
+                    .setOffset(entry.getValue().beforeOffset()));
+
+            System.out.println("Partitions: " + deleteRecords.partitions());
```
Let's remove this debugging statement

mimaison commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,170 @@
+                deleteRecords = new DeleteRecordsRequestData.DeleteRecordsTopic()
+                        .setName(topicPartition.topic());
+                deletionsForTopic.put(topicPartition.topic(), deleteRecords);
+            }
```
Nothing wrong with your logic but we could use `computeIfAbsent` here
```suggestion
            DeleteRecordsRequestData.DeleteRecordsTopic deleteRecords = deletionsForTopic.computeIfAbsent(
                    topicPartition.topic(),
                    key -> new DeleteRecordsRequestData.DeleteRecordsTopic().setName(topicPartition.topic())
            );
```

mimaison reviewed (commented):
Thanks for the PR. I made a first quick pass and left a few comments

tinaselenge committed:
Address review comments

tinaselenge committed:
Remove a test case that is not longer valid
The test was expecting LeaderNotAvailableException when deleting records on non existant partition.
This test case will no longer fail with this exception because now the metadata request will be retried and the delete records request will timeout instead.
Therefore removed this particular test case so that the test does not wait until timeout.

tinaselenge commented:
@divijvaidya @showuon @mimaison Thank you very much for reviewing the PR!

I believe I have addressed the comments now. Please let me know if I have missed anything. Thanks.

showuon commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2925,123 +2919,11 @@ void handleFailure(Throwable throwable) {
     @Override
     public DeleteRecordsResult deleteRecords(final Map<TopicPartition, RecordsToDelete> recordsToDelete,
                                              final DeleteRecordsOptions options) {
+        AdminApiFuture.SimpleAdminApiFuture<TopicPartition, DeletedRecords> future = DeleteRecordsHandler.newFuture(recordsToDelete.keySet());
```
nit: `AdminApiFuture.SimpleAdminApiFuture` -> `SimpleAdminApiFuture` should work.

showuon commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandler.java:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        DeleteRecordsRequestData data = new DeleteRecordsRequestData()
+                .setTopics(new ArrayList<>(deletionsForTopic.values()));
```
Should we set timeout here like before?

tinaselenge replied:
I could be wrong but I think the timeout specified by the user gets passed to AdminApiDriver (via invokeDriver()) and it handles the timeout for the requests.

showuon replied:
No, I don't think they are the same thing. The invokeDriver timeout is set in the client side for client timeout the request. But the `timeoutMs` field in `DeleteRecordsRequest` is to send to the server side to notify the server about the timeout.

REF: https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/DeleteRecordsRequest.json#L39

showuon reviewed (commented):
Thanks for the update. Left some comments.

showuon commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -2359,31 +2334,22 @@ public void testDeleteRecords() throws Exception {
+            // not authorized to delete records for partition 2
             KafkaFuture<DeletedRecords> myTopicPartition2Result = values.get(myTopicPartition2);
             try {
                 myTopicPartition2Result.get();
                 fail("get() should throw ExecutionException");
             } catch (ExecutionException e1) {
-                assertTrue(e1.getCause() instanceof LeaderNotAvailableException);
+                assertTrue(e1.getCause() instanceof TopicAuthorizationException);
             }
```
This can be replaced with `assertThrows(TopicAuthorizationException.class, () ->myTopicPartition2Result.get(), "get() should throw ExecutionException" )`

Also, Should we change the comment into `TopicAuthorizationException`? It doesn't throw `ExecutionException` at all.

Same comment applies to other places.

tinaselenge replied:
We are actually catching ExecutionException and checking the cause of the exception. This would not work because it's not throwing TopicAuthorizationException, but it is the supplied cause for the ExecutionException.

showuon replied:
Ah, I see. Thanks for the explanation.

showuon commented on clients/src/test/java/org/apache/kafka/clients/admin/internals/DeleteRecordsHandlerTest.java:
```diff
@@ -0,0 +1,209 @@
+        failed.put(errorPartition, error.exception());
+        Set<TopicPartition> completed = new HashSet<>(recordsToDelete.keySet());
+        completed.removeAll(failed.keySet());
+        assertResult(result, completed, failed, emptyList(), emptySet());
```
Could we add a test to verify a mixed case? I.e. a test case contains completed, failed, unmapped, and retriable.

tinaselenge committed:
Address review comments

showuon commented:
@tinaselenge , it looks like some tests are failed: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13760/6/
Please also take a look. Thanks.

tinaselenge committed:
fix unit test failure

tinaselenge committed:
Implement timeout for DeleteRecords request

tinaselenge commented:
Thank you @showuon  for the feedback. I think I have addressed the comments. Please let me know there is anything I missed or doesn't look right.

showuon reviewed (approved):
LGTM!

showuon commented:
Re-running CI build: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13760/9/

showuon commented:
Failed tests are unrelated.
